stock_available_product_expiry/models/product_product.py

```diff
@@ -0,0 +1,179 @@
+            product.qty_expired = expired_quantities['qty_expired']
+            product.outgoing_expired_qty = expired_quantities[
+                'outgoing_expired_qty']
+            if not self.env.context.get('disable_check_expired_lots', False):
+                # We let this to have possibility to get standard quantity
+                product.outgoing_qty -= product.outgoing_expired_qty
+            product.qty_expired -= product.outgoing_expired_qty
+        return res
+
+    @api.model
+    @contextmanager
+    def _force_auto_join(self, model_field_list):
+        """ Force  auto_join for each item (model_name, field_name)
+        in model_field_list
+        The original value are restored on exit
+        """
+        vals = {}
+        for model_name, field_name in model_field_list:
+            field = self.env[model_name]._fields[field_name]
+            vals[field] = field.auto_join
+            field.auto_join = True
+        try:
+            yield
+        finally:
+            for field, value in vals.iteritems():
+                field.auto_join = value
+
+    def _get_quants(self, quant_domain, quant_out_domain):
+        # Get auto_join value and reset them after read groups
+        # This to improve performances without breaking normal behaviour
+        stock_quant_obj = self.env['stock.quant']
+        with self._force_auto_join([
+            ('stock.quant', 'lot_id'),
+                ('stock.quant', 'reservation_id')]):
+            quants = stock_quant_obj.with_context(lang='').read_group(
+                quant_domain, ['product_id', 'qty'], ['product_id'])
+            quants_out = stock_quant_obj.with_context(lang='').read_group(
+                quant_out_domain, ['product_id', 'qty'], ['product_id'])
+            return quants, quants_out
+
+    @api.multi
+    def _compute_expired_quants_dict(self):
+        quant_domain, quant_out_domain = self._get_expired_quants_domain()
+        quants, quants_out = self._get_quants(quant_domain, quant_out_domain)
+        quant_res = {
+            item['product_id'][0]: item['qty'] for item in quants}
+        quant_out_res = {
+            item['product_id'][0]: item['qty'] for item in quants_out}
+        res = dict()
+        for product in self.with_context(prefetch_fields=False):
+            quantities = {}
+            res[product.id] = quantities
+            quantities['outgoing_expired_qty'] = float_round(
+                quant_out_res.get(product.id, 0.0),
+                precision_rounding=product.uom_id.rounding
+            )
+            quantities['qty_expired'] = float_round(
+                quant_res.get(product.id, 0.0),
+                precision_rounding=product.uom_id.rounding
+            )
+        return res
+
+    @api.multi
+    def action_open_expired_quants(self):
+        action = self.env.ref('stock.product_open_quants').read()[0]
+        domain = [
+            ('product_id', '=', self.id),
```

Comment:
maybe change to ('product_id', 'in', self.ids) and this can get called from a list view